pkg/migrator/migrator.go

```diff
@@ -0,0 +1,251 @@
+// DataStore CatalogSources) are directly created by the OperatorSources.
+// The migrator updates existing Subscriptions that referenced Installed CatalogSources
+// during operator installation in an Openshift 4.1.x cluster, to reference
+// Datastore CatalogSources instead. The migrator also deletes the stale
+// CatalogSourceConfigs that were created during operator installation, and
+// the datastore CatalogSourceConfigs created by OperatorSources
+// in an Openshift 4.1.x cluster.
+func NewMigrator(client client.Client) migrator {
+	return migrator{
+		logger: log.NewEntry(log.New()),
+		client: client,
+	}
+}
+
+type migrator struct {
+	logger *logrus.Entry
+	client client.Client
+}
+
+// Migrate updates existing Subscriptions, deletes the CSCs installed during
+// operator installation as well the stale datastore CSCs when the cluster is
+// migrating from Openshift 4.1.z to Openshift 4.2.z
+func (m *migrator) Migrate(operatorNamespace string) error {
+	installedCscs, errors := m.updateSubscriptions()
+	errors = append(errors, m.deleteInstalledCscs(installedCscs, operatorNamespace)...)
+	errors = append(errors, m.deleteDatastoreCscs(operatorNamespace)...)
+	if errors != nil {
+		return fmt.Errorf("[migration] Failed to complete migration")
+	}
+	return nil
+}
+
+// updateSubscriptions updates the existing Subscriptions'
+// spec.source and spec.sourcenamespace fields. Existing
+// Subscriptions referenced installed CatalogSources, which
+// are updated to reference datastore CatalogSources
+// instead.
+func (m *migrator) updateSubscriptions() ([]types.NamespacedName, []error) {
+	var installedCscs []types.NamespacedName
+	var errors []error
+	options := &client.ListOptions{}
+	options.SetLabelSelector(fmt.Sprintf(builders.CscOwnerNameLabel))
+
+	subscriptions := &olm.SubscriptionList{}
+	// Get the list of existing Subscriptions that have the label "csc-owner-name"
+	err := m.client.List(context.TODO(), options, subscriptions)
+	if err != nil {
+		msg := fmt.Sprintf("[migration] Client error: %s", err.Error())
+		errors = append(errors, fmt.Errorf(msg))
+		m.logger.Errorf(msg)
+		return []types.NamespacedName{}, errors
+	}
+	for _, subscription := range subscriptions.Items {
+		if builders.HasOwnerLabels(subscription.GetLabels(), v2.CatalogSourceConfigKind) {
+			installedCscs = append(
+				installedCscs,
+				types.NamespacedName{
+					Name:      subscription.GetLabels()[builders.CscOwnerNameLabel],
+					Namespace: subscription.GetLabels()[builders.CscOwnerNamespaceLabel]})
+			// Try to infer the datastore CatalogSource from the Subscription
+			datastoreCs, err := findCatalogSource(&subscription, m.client)
+			if k8s_errors.IsNotFound(err) {
+				// Infer the CatalogSource from the OperatorSource that has the package
+				datastoreCs, err = findCsFromOpsrc(&subscription, m.client)
+				if err != nil {
+					msg := fmt.Sprintf("[migration] Could not infer datastore CatalogSource for Subscription %s", subscription.GetName())
```

Comment:
Why are you throwing away the error that you got and replacing it with a custom error? It is better to propagate that error up instead of replacing it with your own error. Adding a message to the existing error is fine like you are doing other places. This comment applies to all places that you are doing this. This applies to all places that you are doing this.

Reply:
I am not sure I understand. We are raising these errors in the log as they appear. I am only surfacing these errors so that I know whether the migration failed or not and displaying that it failed. I get that we are not using these surfaced errors for anything right now and so collecting them might be overkill for the current usage. Is there a particular way you imagined these errors being used?

Reply:
Where is `err` not being reflected in the message you are constructing? What is the point in getting a log from a customer with the error message that you have shown? You are anyhow collecting the error messages. So why throw away the real error when you are collecting them?

Reply:
Oh ok I didn't notice `err` wasn't being reported back in that error! Yes I totally agree - that's my bad. Sorry for the misunderstanding